// @date-vir/duration/dist/units/date-unit.js
/**
 * All date units supported by date-vir. These are the singular versions of {@link DurationUnit}.
 *
 * @category Unit
 */
export var DateUnit;
(function (DateUnit) {
    DateUnit["Year"] = "year";
    DateUnit["Quarter"] = "quarter";
    DateUnit["Month"] = "month";
    DateUnit["Week"] = "week";
    DateUnit["Day"] = "day";
    DateUnit["Hour"] = "hour";
    DateUnit["Minute"] = "minute";
    DateUnit["Second"] = "second";
    DateUnit["Millisecond"] = "millisecond";
})(DateUnit || (DateUnit = {}));
/**
 * All {@link DateUnit} values that are zero indexed (start at 0, like seconds).
 *
 * @category Internal
 */
export const zeroIndexedDateUnits = [
    DateUnit.Year,
    DateUnit.Hour,
    DateUnit.Minute,
    DateUnit.Second,
    DateUnit.Millisecond,
];
/**
 * All {@link DateUnit} values that are one indexed (start at 1, like days of a month).
 *
 * @category Internal
 */
export const oneIndexedDateUnits = [
    DateUnit.Quarter,
    DateUnit.Month,
    DateUnit.Week,
    DateUnit.Day,
];
/**
 * All {@link DateUnit} values in an array ordered from smallest to largest.
 *
 * @category Internal
 */
export const orderedDateUnit = [
    DateUnit.Millisecond,
    DateUnit.Second,
    DateUnit.Minute,
    DateUnit.Hour,
    DateUnit.Day,
    DateUnit.Week,
    DateUnit.Month,
    DateUnit.Quarter,
    DateUnit.Year,
];
