// @date-vir/duration/dist/units/day-of-week.js
/**
 * The names of each day of the week, in English and capitalized.
 *
 * @category Unit
 */
export var DayOfWeekName;
(function (DayOfWeekName) {
    DayOfWeekName["Sunday"] = "Sunday";
    DayOfWeekName["Monday"] = "Monday";
    DayOfWeekName["Tuesday"] = "Tuesday";
    DayOfWeekName["Wednesday"] = "Wednesday";
    DayOfWeekName["Thursday"] = "Thursday";
    DayOfWeekName["Friday"] = "Friday";
    DayOfWeekName["Saturday"] = "Saturday";
})(DayOfWeekName || (DayOfWeekName = {}));
/**
 * Bounds for valid day of the week indexes.
 *
 * @category Internal
 */
export const dayOfWeekIndexBounds = {
    min: 0,
    max: 6,
};
/**
 * Maps {@link DayOfWeekName} to {@link DayOfWeekIndex}.
 *
 * @category Util
 */
export const dayOfWeekNameToIndex = {
    [DayOfWeekName.Sunday]: 0,
    [DayOfWeekName.Monday]: 1,
    [DayOfWeekName.Tuesday]: 2,
    [DayOfWeekName.Wednesday]: 3,
    [DayOfWeekName.Thursday]: 4,
    [DayOfWeekName.Friday]: 5,
    [DayOfWeekName.Saturday]: 6,
};
/**
 * An array of {@link DayOfWeekName} in index order. This can be accessed by any
 * {@link DayOfWeekIndex} number.
 *
 * @category Util
 */
export const dayOfWeekNameOrder = [
    DayOfWeekName.Sunday,
    DayOfWeekName.Monday,
    DayOfWeekName.Tuesday,
    DayOfWeekName.Wednesday,
    DayOfWeekName.Thursday,
    DayOfWeekName.Friday,
    DayOfWeekName.Saturday,
];
