// @date-vir/duration/dist/units/month.js
/**
 * Names of all months in English.
 *
 * @category Unit
 */
export var MonthName;
(function (MonthName) {
    MonthName["January"] = "January";
    MonthName["February"] = "February";
    MonthName["March"] = "March";
    MonthName["April"] = "April";
    MonthName["May"] = "May";
    MonthName["June"] = "June";
    MonthName["July"] = "July";
    MonthName["August"] = "August";
    MonthName["September"] = "September";
    MonthName["October"] = "October";
    MonthName["November"] = "November";
    MonthName["December"] = "December";
})(MonthName || (MonthName = {}));
/**
 * All English month names in order.
 *
 * @category Unit
 */
export const orderedMonthNames = [
    MonthName.January,
    MonthName.February,
    MonthName.March,
    MonthName.April,
    MonthName.May,
    MonthName.June,
    MonthName.July,
    MonthName.August,
    MonthName.September,
    MonthName.October,
    MonthName.November,
    MonthName.December,
];
/**
 * Bounds for valid month numbers.
 *
 * @category Internal
 */
export const monthNumberBounds = {
    min: 1,
    max: 12,
};
/**
 * Bounds for valid day of the month numbers.
 *
 * @category Internal
 */
export const dayOfMonthBounds = {
    min: 1,
    max: 31,
};
